Fixed buggy interactions between button presses and focus changes

e8newallm opened the pull request:
Fix #52578

This bug was caused by the button_up event being sent to the newly focused control rather than the original. With this bug, I also discovered that using LMB, enter, and space, you could trigger multiple button_down/pressed events with no button_up in between them.

The multiple button downs bug has been fixed by adding a status.button_down flag which tracks whether the button is pressed down or not, and not permitting the button to be pressed down further when already pressed or releasing further when it's unpressed.

The original issue proposed having the original button receive the button_up event even when the focus has changed, however this feels like it would take a large restructuring of the code to achieve.

I have set it up so when a button is currently down, and the focus is changed, the button is unpressed with no pressed event if the button is in Button Release mode. I see this as the best solution to this problem, as to still pass key-presses to an unfocused button seems unintuitive to me. It also serves as a way to cancel an accidental button press. I would also argue the same should be implemented for the mouse presses too (When the mouse no longer hovers over the button, the button down should be undone).

Another possible addition to this could be to button up the now unfocused button, and button down the newly focused button, however this could cause awkwardness with a mix of Button Release and Button Press buttons.

I am open to changing this based on criticism.

Calinou commented:
Note: Since this results in a behavior change, I think this should only be cherry-picked to `3.x`, not `3.3`.

e8newallm committed:
Fixed bug where you could trigger multiple button_down and pressed signals without a button_up between
Fixed bug where changing focus while button is held down triggers a button up on the wrong button

akien-mga commented on scene/gui/base_button.cpp:
```diff
@@ -133,9 +138,11 @@ void BaseButton::_toggled(bool p_pressed) {
 void BaseButton::on_action_event(Ref<InputEvent> p_event) {
-	if (p_event->is_pressed()) {
+	if (p_event->is_pressed() && !status.pressed_down) {
+
```
Style issue, there shouldn't be this empty line after the opening brace.

akien-mga commented:
I feel like we've been tweaking this code countless times over the years, often with weird regressions, so I'm a bit cautious by default here. But if @KoBeWi thinks it's good I think we can give it a spin, could you rebase and retest to make sure this still behaves as intended @e8newallm?

KoBeWi commented:
I already tested the PR and it basically changes 2 things:
- if a button is being pressed and you change button's focus using keyboard, it will be unpressed
- when you press the button multiple times without releasing it (using mouse, keyboard etc.), only one button down signal will be emitted. Same the other way, i.e. you can only release a pressed button once

To me this change makes sense, but maybe there was some obscure code that relied on the old behavior (chances are low though).

The only 2 lines I'm not sure about are these:
![image](https://user-images.githubusercontent.com/2223172/152253220-62cec129-38fe-41cd-a535-7a06390b381f.png)
(idk how to make a permalink)
Doesn't look harmful, but also doesn't have visible effect, so it'd be nice to know their purpose.

Calinou commented:
>  (idk how to make a permalink)

Select lines on GitHub, press <kbd>Y</kbd> and copy the URL in your address bar.

pycbouh commented:
> maybe there was some obscure code that relied on the old behavior

Maybe CheckBox/CheckButton can be affected, or just toggled buttons in general.

KoBeWi commented:
I checked and they don't seem to have problems either.

>Select lines on GitHub, press Y and copy the URL in your address bar. This works on any file view on GitHub 🙂

https://github.com/godotengine/godot/pull/52621/commits/505815aaeb8c159da4967a7349ff232f551c0894#diff-019e9a782fb47f26ffd4b60b130a23d0b222442a2755917dc31131c3023670c2R169-R170

Like this? I figured that before, but it doesn't expand to a nice preview like with regular permalinks.

KoBeWi commented:
I was checking another PR and it inspired me to check one more thing. And there is a regression here actually:
![godot windows tools 64_4V8trbj7kl](https://user-images.githubusercontent.com/2223172/152340243-ed672d09-7d8a-46a5-9260-c6816e5cb42d.gif)
The button to the left has `keep_pressed_outside` set to true. When you press it and move cursor outside without releasing and then release it outside, the button stays pressed.

The one to the right is even more bugged. It doesn't have `keep_pressed_outside`, so it gets "unpressed" when you move cursor outside. But after you release the button outside and move cursor back above button, it gets pressed for no reason.

So this needs fixing.
